Room airflow network

lgu1234 opened the pull request:
@jasondegraw The branched was committed. After commit, merge was performed. Then a pull request was created. Please review the code.

lgu1234 committed:
New feature code check-in: Room Air Model using Airflow Network

Added a new module: RoomAirModelAirflowNetwork
Added a new example input file: RoomAirflowNetwork.idf
No differences were found by design day switch. Performed annual run

No differences were found by running all example files on a local machine.

lgu1234 committed:
Merge branch 'develop' into RoomAirflowNetwork

Conflicts:
	doc/src/docs/InputOutputReference/01a-InputOutputReference.md
	src/EnergyPlus/DataAirflowNetwork.hh
	src/EnergyPlus/ZonePlenum.cc
	tst/EnergyPlus/unit/CMakeLists.txt

Myoldmopar commented:
@lgu1234 Here are some first-pass thoughts from me:
- The code won't compile _most likely_ because of [this](https://github.com/NREL/EnergyPlus/pull/5081/files#diff-6f7a5dbef69ca9c0c3644f9be817ed32R2345) line.  You are calling `ma` with an argument that hasn't been put into ObjexxFCL.  This is what we are trying to clean up in #5056.  Since there are other issues in the branch, I would hold off on changing anything just yet.
- Was your intention with [this table](https://github.com/NREL/EnergyPlus/pull/5081/files#diff-c60eaa266a85562444de29fcc15542d9R76) to have all those entries in the same cell?
- [This](https://github.com/NREL/EnergyPlus/pull/5081/files#diff-c60eaa266a85562444de29fcc15542d9R1) html section needs to be removed before your commit.  This is already inserted when the documentation is built.
- Something went wrong with [this](https://github.com/NREL/EnergyPlus/pull/5081/files#diff-fb6ff352da9f540fb335f679c2e80bb6) documentation file.  It shows every file changed.  Typically a line ending problem.  I would probably just checkout a fresh version and re-apply your changes being careful to convert all to Windows line endings before committing so that you don't end up with a mixed line endings problem.
- [Copyright years](https://github.com/NREL/EnergyPlus/pull/5081/files#diff-3e2b2fb896a4cabf6636474e423e1825R78) need to be 2015

I'll have a look at the actual implementation code soon, but try to get these side issues worked out.

(FYI @jasondegraw)

DeadParrot commented:
A few comments:
- I believe the `ma` use with `SurfaceName` is correct: he is bypassing the pre-canned names in ObjexxFCL by using the `ma` call directly.
- The compile failures on Linux are, I think, due to the use of `IntraZoneNodeData.Name()` because `Name` is a member of the base class and template arg deduction can't do the conversion. (VC seems to do the conversion.) I sent @lgu1234 the work-around using a reference to the base class.
- As @Myoldmopar  hints, we want to move away from member array usage and the outstanding PR #5056 has a better way of doing these so someone will need to update the usage in this branch before it is merged into develop (hopefully by then #5056 has been merged).
- If `AirflowNetworkLinkage` is an abstract base class it needs to be made proper and non-instantiable (protected constructors, etc.). If not the code should be refactored to avoid deriving from a concrete class. In either case the base needs a public virtual destructor and it is probably good form to put those in all classes in the hierarchy.

lgu1234 commented:
@Myoldmopar 

Here are my answers and actions to your concerns.

1 Code compilation

I removed the base struct in the struct IntraZoneLinkageProp and make it standalone. My purpose is to try it to see what happens. I know this is not a final solution. We can finalize it later. Unfortunately, I don't have Linux machine to test. That is why I commit it and let CI machine tell me what happens.

2 Table items

This is my intention to use consistent style with other table items.

3 This html section

I removed the inserted section.

4 Something wrong in the documentation

I think the inline equation may be wrong. However, I don't know how to make it right, because I don't have a right tool to see. Please advice.

5 Copyright (c) 1996-2014

It should be 2015. I find 2014 happens every file. Should I make changes?

lgu1234 committed:
Revised struct IntraZoneLinkageProp for testing
Removed an html section in the beginning of 12-AlternativeModeling.md

Myoldmopar commented:
@lgu1234 Thanks for the response.
- As for the documentation being wrong, I wasn't talking about a particular equation or line, I was saying that the diff shows that _every_ line of the file has changed.  This is commonly because something happens with the line endings on the file and it gets confused.  Don't worry about it right now, but we'll make sure to resolve it before you merge.
- As for the copyright year, gah, I had a pull request a few weeks back that was supposed to fix that.  It didn't.  I have a pull request right now that updates the copyright year on the existing EnergyPlus source.  When #5085 gets merged into develop, please:
  - pull develop into your branch to get your branch up to date, and
  - update the year to 2015 in any _new_ source files you add with this branch.

lgu1234 commented:
@Myoldmopar Thanks for the update. I will leave documentation right now and will make copyright date change when a new pull request is created.

lgu1234 committed:
Remove struct AirflowNetworkLinkage for diagnosis

jasondegraw commented on idd/Energy+.idd.in:
```diff
@@ -1,5 +1,5 @@
-!IDD_Version ${CMAKE_VERSION_MAJOR}.${CMAKE_VERSION_MINOR}.${CMAKE_VERSION_PATCH}
-!IDD_BUILD ${CMAKE_VERSION_BUILD}
+!IDD_Version 8.3.3
+!IDD_BUILD Unknown
```
@lgu1234 I'm pretty sure these two lines and the next one below need to be reverted back

lgu1234 replied:
@jasondegraw Thanks for pointing this out. I reverted it back in Energy+.idd.in on my local machine and commit it soon.

lgu1234 committed:
Cleanup by removing unused variables and revert top two line of idd

mjwitte commented on idd/Energy+.idd.in:
```diff
@@ -372,7 +372,7 @@ Version,
   A1 ; \field Version Identifier
       \required-field
-      \default ${CMAKE_VERSION_MAJOR}.${CMAKE_VERSION_MINOR}
+      \default 8.3
```
@lgu1234 This line also needs to be reverted.

lgu1234 replied:
@mjwitte Thanks. I reverted it on my local computer and will make a commit with other potential changes.

mjwitte commented on idd/Energy+.idd.in:
```diff
@@ -18385,6 +18385,9 @@ RoomAirModelType,
       \note (2-Node UFAD model for exterior zones)
       \note needs RoomAirSettings:UnderFloorAirDistributionExterior object referencing this Zone
+      \key AirflowNetwork
+      \note (RoomAir modeling using AirflowNetwork)
```
@lgu1234 Notes are at the field level, not the key level, so this should be something like "AirflowNetwork = RoomAir modeling using AirflowNetwork.  Several other notes in this field need a similar change.

lgu1234 replied:
@mjwitte OK. I changed and made other notes changed.

mjwitte commented on idd/Energy+.idd.in:
```diff
@@ -19197,6 +19200,408 @@ RoomAirSettings:UnderFloorAirDistributionExterior,
+  A1, \field Name
+      \type Alpha
+      \reference RoomAirNodeGains
+  A2, \field Internal Gain Object 1 Type
```
@lgu1234 I know it is late in the game to say this, but this object will be difficult to maintain and difficult to use. When a new internal gain object is added, the developer of that object likely won't know that it needs to be added here - both in code and IDD.  @Myoldmopar Perhaps we need to start a list of dependencies in the developer guide?

lgu1234 replied:
@mjwitte So, I am going to keep as is for the time being.

mjwitte commented on idd/Energy+.idd.in:
```diff
@@ -19197,6 +19200,408 @@ RoomAirSettings:UnderFloorAirDistributionExterior,
+      \key PlantComponent:UserDefined
+      \key Coil:UserDefined
+      \key ZoneHVAC:ForcedAir:UserDefined
+      \key AirTerminal:SingleDuct:UserDefined  
```
@lgu1234 Why are the UserDefined components here and not in the HVAC list in the next object?  Or do these need to be in both places?

lgu1234 replied:
@mjwitte The UserDefined components should not be used in both places. Otherwise, it will cause double counting. Since the components were listed before I took over, I just uses as is.

lgu1234 committed:
Revert DataAirflowNetwork.hh with base struct
Made minor changes of idd based on Mike's comments

jasondegraw committed:
Merge remote-tracking branch 'upstream/develop' into RoomAirflowNetwork

jasondegraw committed:
Move to new FindItem and VerifyName in RAFN

jasondegraw commented on src/EnergyPlus/DataAirflowNetwork.hh:
```diff
@@ -1021,6 +1031,110 @@ namespace DataAirflowNetwork {
+			Height( Height ),
+			RAFNNodeNum( RAFNNodeNum ),
+			ZoneNum( ZoneNum ),
+			AFNZoneNum( AFNZoneNum )
```
@lgu1234 Is there an argument missing here? This initializes AFNZoneNum with itself.

lgu1234 replied:
@jasondegraw You are right. I am going to add it below line 1058. Thanks.

jasondegraw committed:
Address some of the Linux warnings

jasondegraw commented:
@lgu1234 I've addressed some of the Linux warnings, but there are still a few more to get worked out.

lgu1234 commented:
@jasondegraw I am going to address reamaing warnings:

AFNZoneNum, TempTmp and HumRatTmp

jasondegraw commented on src/EnergyPlus/RoomAirModelAirflowNetwork.cc:
```diff
@@ -0,0 +1,1487 @@
+		TempIndCoef = ThisRAFNNode.SumIntSensibleGain + ThisRAFNNode.SumHATsurf - ThisRAFNNode.SumHATref + ThisRAFNNode.SumLinkMCpT + ThisRAFNNode.SumSysMCpT + ThisRAFNNode.NonAirSystemResponse + ThisRAFNNode.SysDepZoneLoadsLagged;
+		AirCap = ThisRAFNNode.AirVolume * ZoneVolCapMultpSens * ThisRAFNNode.RhoAir * ThisRAFNNode.CpAir / ( TimeStepSys*SecInHour );
+
+		{ auto const SELECT_CASE_var( ZoneAirSolutionAlgo );
```
@lgu1234 The warning about TempTmp is in here. We should change this into a switch statement with a default. Is there a default value for ZoneAirSolutionAlgo?

lgu1234 replied:
Yes. The default vale is Use3rdOrder. I am thinking to initialize TempTmp in definition, such as 

int TempTmp(0.0);

lgu1234 replied:
@jasondegraw I am going to make a switch statement.

jasondegraw commented on src/EnergyPlus/RoomAirModelAirflowNetwork.cc:
```diff
@@ -0,0 +1,1487 @@
+		B = ( ThisRAFNNode.SumIntLatentGain / H2OHtOfVap ) + ThisRAFNNode.SumSysMW + ThisRAFNNode.SumLinkMW + ThisRAFNNode.SumHmARaW;
+		C = ThisRAFNNode.RhoAir * ThisRAFNNode.AirVolume * ZoneVolCapMultpMoist / ( SecInHour * TimeStepSys );
+
+		{ auto const SELECT_CASE_var( ZoneAirSolutionAlgo );
```
@lgu1234 This should also be a switch statement.

lgu1234 replied:
Same as before.

jasondegraw commented on src/EnergyPlus/RoomAirModelAirflowNetwork.cc:
```diff
@@ -0,0 +1,1487 @@
+			HA = HA + HConvIn( SurfNum ) * Area;
+			SumHATsurf += HConvIn( SurfNum ) * Area * TempSurfInTmp( SurfNum );
+
+			{	auto const SELECT_CASE_var( Surface( SurfNum ).TAirRef );
```
@lgu1234 Another candidate for a switch statement

lgu1234 replied:
@jasondegraw I am going to address it.

lgu1234 committed:
Address remaining Linux warnings

jasondegraw commented:
@lgu1234 There was something going on with the CI, but I thought it was fixed.

@Myoldmopar Is there still something going on?

lgu1234 committed:
Replace a section of SELECT_CASE_var by IF statements

jasondegraw commented on src/EnergyPlus/RoomAirModelAirflowNetwork.cc:
```diff
@@ -0,0 +1,1482 @@
+
+		// SUBROUTINE INFORMATION:
+		//       AUTHOR         Brent Griffith
+		//       DATE WRITTEN   January 2004/Aug 2005
```
@lgu1234 I'm pretty sure that this subroutine information comment is out of date, so it either needs to be revised or removed. The same goes for all of the other info comments. It is fine with me if you just remove all of these comments from the new functions. The history is all contained in git, so... I guess I would remove everything but the "purpose" comment.

But you and/or others may have a different opinion, so I'll leave it up to you (@mjwitte @Myoldmopar ping).

mjwitte replied:
@jasondegraw Well, that would be something that the programming standard should cover.  These comment blocks were required.  Relying on github history will be lost when the day comes that github is no longer and we move to something else.  So, I don't know if we've made any conscious decision on changing the standard. The sooner we clarify these things, the better.

DeadParrot replied:
@mjwitte I agree that this should get resolved soon. The Fortran comment blocks are mostly cruft now and all the "na" lines are silly. The history held by git (not GitHub) is good for the "who" and "when" but may not give a full picture of the "why", so a short section with the author and major revision info makes sense. Anything that the code tells you (callers, callees, ...) should not be in the comments because they are always out of date and other tools can extract that from the code.

It is worth considering tagged comments for a system like doxygen -- not a big fan of that but if we want a hyperlinked API some day those tags help.

I have started work on a [coding guidelines](https://github.com/NREL/EnergyPlus/wiki/Coding-Guidelines) draft as discussed at LBL and a guide for class and function commenting should (eventually) be part of that. I'm willing to put a suggested commenting banner guideline in there so that there is even more for everyone to ~~hate on~~ contribute to.

lgu1234 replied:
@jasondegraw I am conservative and agree with @mjwitte opinion. Although GitHub keeps history, it is also good to see history directly from the code.

jasondegraw replied:
@lgu1234 It seems like the best thing to do for now is to just revise the info comments. You guys can decide on what comments need to be there - the "why" is the only one that I'd require... but for once I have no opinion either way

jasondegraw commented:
@lgu1234 I think most of those warnings are due to "using" statements when the variable isn't actually used. In CheckEquipName in RoomAirModelManager.cc:

```
...
using DataLoopNode::NumOfNodes;
using DataLoopNode::NodeID;
using DataLoopNode::Node;
...
```

NodeID is used but the other two are not.

lgu1234 committed:
Remove unused using statements

lgu1234 committed:
Remove the last unused using statement

lgu1234 committed:
Merge branch 'develop' into RoomAirflowNetwork

jasondegraw commented:
@lgu1234 I'll do a more detailed code review next. It looks like the info comments haven't been updated. Could you do that next?

lgu1234 commented:
@jasondegraw Could you be more specific? My understanding is to remove following comments in all functions in the RoomAirModelAirflowNetwork module?

```
    // SUBROUTINE INFORMATION:
    //       AUTHOR         B. Griffith
    //       DATE WRITTEN   November 2009
    //       MODIFIED       na
    //       RE-ENGINEERED  na

    // PURPOSE OF THIS SUBROUTINE:
    // <description>

    // METHODOLOGY EMPLOYED:
    // <description>

    // REFERENCES:
    // na
```

Could you confirm?

jasondegraw commented:
@lgu1234 I think the conclusion of the earlier discussion was to keep the comments. Based upon that conversation, I'd say we need to make sure that the authorship info and descriptive info (purpose and/or methodology) are correct.

lgu1234 commented:
@jasondegraw I am going to check information on authorship and description to ensure they are updated. Hope you will complete detailed code review soon.

jasondegraw commented on src/EnergyPlus/RoomAirModelManager.cc:
```diff
@@ -1350,6 +1365,463 @@ namespace RoomAirModelManager {
+			for ( ZoneNum = 1; ZoneNum <= NumOfZones; ++ZoneNum ) {
+				// find surface list
+				if ( RoomAirflowNetworkZoneInfo( ZoneNum ).NumOfAirNodes > 0 ) {
+					RAFNNodeNum = FindItemInList( cAlphaArgs( 1 ), RoomAirflowNetworkZoneInfo( ZoneNum ).Node.ma( &RoomAirflowNetworkAirNodeNestedStruct::NodeSurfListName ), RoomAirflowNetworkZoneInfo( ZoneNum ).NumOfAirNodes );
```
@lgu1234 There should be a new version of FindItemInList that doesn't require calling ma directly, which we should avoid. I think this is it:

https://github.com/NREL/EnergyPlus/blob/develop/src/EnergyPlus/InputProcessor.hh#L667

lgu1234 replied:
@

lgu1234 replied:
@jasondegraw Corrected as
                    RAFNNodeNum = FindItemInList( cAlphaArgs( 1 ), RoomAirflowNetworkZoneInfo( ZoneNum ).Node, &RoomAirflowNetworkAirNodeNestedStruct::NodeSurfListName, RoomAirflowNetworkZoneInfo( ZoneNum ).NumOfAirNodes );

jasondegraw commented on src/EnergyPlus/RoomAirModelManager.cc:
```diff
@@ -1350,6 +1365,463 @@ namespace RoomAirModelManager {
+			for ( ZoneNum = 1; ZoneNum <= NumOfZones; ++ZoneNum ) {
+				// find surface list
+				if ( RoomAirflowNetworkZoneInfo( ZoneNum ).NumOfAirNodes > 0 ) {
+					RAFNNodeNum = FindItemInList( cAlphaArgs( 1 ), RoomAirflowNetworkZoneInfo( ZoneNum ).Node.ma( &RoomAirflowNetworkAirNodeNestedStruct::NodeIntGainsListName ), RoomAirflowNetworkZoneInfo( ZoneNum ).NumOfAirNodes );
```
@lgu1234 New FindItemInList should be used instead of ma call

lgu1234 replied:
@jasondegraw Corrected as 
                    RAFNNodeNum = FindItemInList( cAlphaArgs( 1 ), RoomAirflowNetworkZoneInfo( ZoneNum ).Node, &RoomAirflowNetworkAirNodeNestedStruct::NodeIntGainsListName, RoomAirflowNetworkZoneInfo( ZoneNum ).NumOfAirNodes );

jasondegraw commented on src/EnergyPlus/RoomAirModelManager.cc:
```diff
@@ -1350,6 +1365,463 @@ namespace RoomAirModelManager {
+			for ( ZoneNum = 1; ZoneNum <= NumOfZones; ++ZoneNum ) {
+				// find surface list
+				if ( RoomAirflowNetworkZoneInfo( ZoneNum ).NumOfAirNodes > 0 ) {
+					RAFNNodeNum = FindItemInList( cAlphaArgs( 1 ), RoomAirflowNetworkZoneInfo( ZoneNum ).Node.ma( &RoomAirflowNetworkAirNodeNestedStruct::NodeHVACListName ), RoomAirflowNetworkZoneInfo( ZoneNum ).NumOfAirNodes );
```
@lgu1234 This is the last ma call that I found, let's replace this one as well

lgu1234 replied:
@jasondegraw Corrected as
                    RAFNNodeNum = FindItemInList( cAlphaArgs( 1 ), RoomAirflowNetworkZoneInfo( ZoneNum ).Node, &RoomAirflowNetworkAirNodeNestedStruct::NodeHVACListName, RoomAirflowNetworkZoneInfo( ZoneNum ).NumOfAirNodes );

jasondegraw commented on src/EnergyPlus/RoomAirModelAirflowNetwork.cc:
```diff
@@ -0,0 +1,1471 @@
+		Real64 SysOutputProvided;
+		Real64 LatOutputProvided;
+
+		// TODO, FINISH THIS
```
@lgu1234 Is this really not finished?

lgu1234 replied:
@jasondegraw Completed. "FINISH THIS" is removed.

jasondegraw commented on src/EnergyPlus/RoomAirModelAirflowNetwork.cc:
```diff
@@ -0,0 +1,1471 @@
+		Real64 LatOutputProvided;
+		int RoomAirNode;
+
+		// TODO
```
@lgu1234 Does this comment still need to be here?

lgu1234 replied:
@jasondegraw  Removed lines between 1398 and 1409.

lgu1234 committed:
Addressed reviewer's recent comments by revising code and function comments in two modules
Added a new example name in CMakeList.txt

jasondegraw committed:
Break up long lines in 12-AlternativeModeling.md

jasondegraw committed:
Update RoomAir description to include RAFN

jasondegraw committed:
Fix typos in 12-AlternativeModeling.md

jasondegraw committed:
Fix equation problems in RAFN docs

jasondegraw committed:
Merge remote-tracking branch 'upstream/develop' into RoomAirflowNetwork

Myoldmopar committed:
Uniformalized line endings in doc file

jasondegraw commented:
@lgu1234 I was about to merge this when I saw that something funny is going on with 01a-InputOutputReference.md. @Myoldmopar volunteered to fix it (thanks) and will do the final merge.

Myoldmopar commented:
After CI works on e5eca3f, I'll have a look over it and merge.
